server/src/main/java/org/elasticsearch/repositories/RepositoriesStatsArchive.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.repositories;
+
+import org.elasticsearch.common.unit.TimeValue;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.List;
+
+public final class RepositoriesStatsArchive {
+    private final TimeValue retentionPeriod;
+    private final Deque<RepositoryStatsSnapshot> archive = new ArrayDeque<>();
+
+    RepositoriesStatsArchive(TimeValue retentionPeriod) {
+        this.retentionPeriod = retentionPeriod;
+    }
+
+    synchronized void archive(RepositoryStatsSnapshot repositoryStats) {
+        if (repositoryStats.getRepositoryInfo().isStopped() == false) {
+            repositoryStats = repositoryStats.withStoppedRepo();
+        }
+        archive.add(repositoryStats);
```

Comment:
Should we assert that we never have two `RepositoryStatsSnapshot` for the same repository ephemeral id in `archive`?